google/cloud/spanner/numeric.h

```diff
@@ -31,83 +40,93 @@ namespace google {
 namespace cloud {
 namespace spanner {
 GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
-class Numeric;  // defined below
+
+// The supported `Decimal` modes.
+enum class DecimalMode {
```

Comment:
As mentioned below, seems like this should be modeled with traits.  I also believe we will need a similar type for BigQuery, with slightly different precision and fraction part (I think):

https://cloud.google.com/bigquery/docs/reference/standard-sql/data-types#decimal_types

Which argues for refactoring to `google/cloud/internal/`

Reply:
We can certainly extract common bits when the time comes.  One thing to note, however, is the mismatch between [max-total-digits + max-fractional-digits] and [max-integral-digits + max-fractional-digits].